kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -95,44 +97,11 @@ private CompletableFuture<Void> doWrite(ChannelHandlerContext ctx, Object msg, C
                 return CompletableFuture.completedFuture(null);
             }
             var future = stage.toCompletableFuture();
-            var withTimeout = handleDeferredStage(ctx, future, decodedFrame);
-            return withTimeout.whenComplete((requestFilterResult, t) -> {
-                // maybe better to run the whole thing on the netty thread.
-
-                if (t != null) {
-                    LOGGER.warn("{}: Filter{} for {} request ended exceptionally - closing connection",
-                            ctx.channel(), filterDescriptor(), decodedFrame.apiKey(), t);
-                    filterContext.closeConnection();
-                    return;
-                }
-
-                if (requestFilterResult.drop()) {
-                    if (LOGGER.isDebugEnabled()) {
-                        LOGGER.debug("{}: Filter{} drops {} request",
-                                ctx.channel(), filterDescriptor(), decodedFrame.apiKey());
-                    }
-                }
-                else if (requestFilterResult.message() != null) {
-                    if (requestFilterResult.shortCircuitResponse()) {
-                        // this is the short circuit path
-                        var header = requestFilterResult.header() == null ? new ResponseHeaderData() : ((ResponseHeaderData) requestFilterResult.header());
-                        header.setCorrelationId(decodedFrame.correlationId());
-                        filterContext.forwardResponseInternal(header, requestFilterResult.message());
-                    }
-                    else {
-                        var header = requestFilterResult.header() == null ? decodedFrame.header() : requestFilterResult.header();
-                        filterContext.forwardRequestInternal((RequestHeaderData) header, requestFilterResult.message());
-                    }
-                }
-
-                if (requestFilterResult.closeConnection()) {
-                    if (requestFilterResult.message() != null) {
-                        ctx.flush();
-                    }
-                    filterContext.closeConnection();
-                }
-            }).toCompletableFuture().thenApply(filterResult -> null);
-
+            boolean defer = !future.isDone();
+            var maybeDeferred = defer ? handleDeferredStage(ctx, future, decodedFrame) : future;
```

Comment:
I'm not 100% sure we can hoist the `isDone` check out of `handleDeferredStage` as the future could complete between this check and `handleDeferredStage` executing. 

I guess its just an optimisation (we avoid creating the timeout future) so it is probably safe.

Reply:
I believe the intent is to optimise a fast path for filters that return completed futures.  I think the code is meeting that aim.